### PR TITLE
feat!: store action built-ins messages by severity in environment

### LINKED ISSUE #187: $fail should wait for report to be generated before failing the action
**Is your feature request related to a problem? Please describe.**
Right now, the report is only generated when all the program is executed.
When the `$fail` action is part of the program, the report doesn't get to be generated because upon the execution of the built-in the GitHub action fails altogether and the report doesn't get to be generated.

**Describe the solution you'd like**
Perform the execution with the following order:
1. Generate the report
2. Execute the `$fail` action.

**Describe alternatives you've considered**
*None*

**Additional context**
*None*

### PR DESCRIPTION
**Pull request by shay2025**
## Description

This pull request a new parameter in the environment which consists of a storage to the action built-ins messages.
For now, we store the fatal messages which once catched, the program action stops and the GitHub action fails.

Further development involves propagating this parameter to store not only fatal messages but all messages returned by action built-ins (and maybe function built-ins), categorizing them by severity (fatal, error, warning, etc) and add them to the generated report.

## Related issue

Closes #187

## Type of change

Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How was this tested?

Used the pull request https://github.com/explore-dev/didactic-robot/pull/47 for testing.

1. Change the reviewpad.yml file in the https://github.com/explore-dev/didactic-robot locally
2. On the branch of the proposed changes executed the command:
`./main -dry-run=false -github-token=[GITHUB_TOKEN] -pull-request=https://github.com/explore-dev/didactic-robot/pull/47 -reviewpad=../didactic-robot/reviewpad.yml -event-payload=cmd/cli/resources/test_open_pr_event.json`

## Checklist

- [x] I have performed a self-review of my code
- [x] I have ran `task check -f` and have no issues

### CODE REVIEW

File: runner.go [hunk 1]
```diff
@@ -77,6 +77,8 @@ func Run(
 			engine.CollectError(evalEnv, err)
 			return nil, err
 		}
+
+		aladinoInterpreter.CheckForFatal()
```
Comment (ferreiratiago) [inline runner.go:81]: No need for this.

**Review by ferreiratiago — CHANGES_REQUESTED** [review]
As we discussed we should update the Interpreter `Report` function in order to report all messages reported from builtins. And also update the Interpreter `ExecProgram` to return the exit status.

**Comment by shay2025** [comment]
Before merging this pull request, upon approval, review https://github.com/reviewpad/action/pull/22

File: engine/env.go [hunk 2]
```diff
@@ -24,7 +24,7 @@ type Interpreter interface {
 	ProcessLabel(id, name string) error
 	ProcessRule(name, spec string) error
 	EvalExpr(kind, expr string) (bool, error)
-	ExecProgram(program *Program) error
+	ExecProgram(program *Program) (int, error)
```
Comment (ferreiratiago) [inline engine/env.go:27]: I would suggest to create a type for the exit status.

```go
type ExitStatus int

const ExitStatusSuccess = 1
const ExitStatusFailure    = 0
```

**Review by ferreiratiago — CHANGES_REQUESTED** [review]
We need to create a type `ExitStatus` to make it more verbose what's happening.

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -29,7 +29,7 @@ jobs:
       - name: Test
         run: task test
 
-      - name: Convert coverage to lcov
+      - name: Convert coverage to lcovreb
```
[the remark this answers is not in the document]
Comment (shay2025) [reply .github/workflows/build.yml:32]: Sorry this resulted of a previous rebase.

File: engine/env.go (hunk 2 above)
Comment (shay2025) [reply engine/env.go:27]: Here I think you meant the opposite. The success exit status is 0 and the failure exit status is 1.

**Review by ferreiratiago — APPROVED** [review]

**Pull request merged**